mixins/dynamic-resize-mixin.js

```diff
@@ -0,0 +1,350 @@
+				reflect: true
+			},
+			_overflowHidden: {
+				type: Boolean,
+			},
+			_chompIndex: {
+				type: Number,
+			},
+		};
+	}
+	constructor() {
+		super();
+		this._handleResize = this._handleResize.bind(this);
+		this._addEventListeners = this._addEventListeners.bind(this);
+		this._chomp = this._chomp.bind(this);
+		this._getLayoutItems = this._getLayoutItems.bind(this);
+		this._handleSlotChange = this._handleSlotChange.bind(this);
+
+		this._throttledResize = throttle(this._handleResize, 15);
+
+		this._overflowHidden = this._overflowHidden ?? false;
+		this.autoShow = this.autoShow ?? false;
+		this.maxToShow = this.maxToShow ?? -1;
+		this.minToShow = this.minToShow ?? 1;
+		this._mini = this.openerType === 'icon';
+	}
+
+	disconnectedCallback() {
+		super.disconnectedCallback();
+		this._removeEventListeners();
+	}
+
+	async firstUpdated() {
+		super.firstUpdated();
+
+		// selected elements
+		this._buttonSlot = this.shadowRoot.getElementById('buttons');
+		this._buttonSlot.addEventListener('slotchange', this._handleSlotChange);
+
+		this._container = this.shadowRoot.querySelector('.d2l-button-group-container');
+
+		this._availableWidth = this._container.clientWidth;
+
+		this._autoShowClass = 'd2l-button-group-show';
+		this._autoNoShowClass = 'd2l-button-group-no-show';
+
+		// get the items from the button slot
+		this._slotItems = this._getSlotItems();
+		// convert them to layout items (calculate widths)
+		this._layoutItems = this._getLayoutItems(this._slotItems);
+		// convert to dropdown items (for overflow menu)
+		this._dropdownItems = this._slotItems.map((node) => this._convertToDropdownItem(node));
+
+
+		if (this.autoShow) {
+			this._autoDetectBoundaries(this._slotItems);
+		}
+		this._chomp();
+
+	}
+	update(changedProperties) {
+		super.update(changedProperties);
+		if (changedProperties.get('autoShow')) {
+			this._autoDetectBoundaries(this._getLayoutItems());
+		}
+
+		if (changedProperties.get('minToShow') || changedProperties.get('maxToShow')) {
+			this._chomp();
+		}
+
+		// slight hack, overflowMenu isnt being rendered initially so wait until update
+		// to calculate its width
+		if (!this._overflowMenuWidth) {
+			// this action needs to be deferred until first render of our overflow button
+			requestAnimationFrame(() => {
+				this._chomp();
+			});
+		}
+	}
+	connectedCallack() {
+		this._addEventListeners();
+	}
+	_addEventListeners() {
+		window.addEventListener('resize', this._throttledResize);
+	}
+	_autoDetectBoundaries(items) {
+
+		let minToShow, maxToShow;
+		for (let i = 0; i < items.length; i++) {
+			if (items[i].classList.contains(this._autoShowClass)) {
+				minToShow = i + 1;
+			}
+			if (maxToShow === undefined && items[i].classList.contains(this._autoNoShowClass)) {
+				maxToShow = i;
+			}
+		}
+
+		if (minToShow !== undefined) {
+			this.minToShow = minToShow;
+		}
+		if (maxToShow !== undefined) {
+			this.maxToShow = maxToShow;
+		}
+	}
+	_chomp(items = this._layoutItems) {
+
```

Comment:
The more I look at this mixin, the less sure I am that making it so generic is that useful. For it to work properly, it needs the things inside it to be pretty specific... buttons and stuff. I'm just not sure it'll be that useable beyond butotn-groups.

The fact that it's a mixin also means all this chomping needs to happen outside of the component's `render()` method, which isn't ideal. If all this code was in `button-group` itself, properties could just be updated, which would cause a re-render and the "chomp" would happen there -- so much simpler. 

@dbatiste curious about your thoughts here? We were originally thinking about the navbar and HTML editor "chompers" when hoping to make this generic.